### ISSUE TITLE
new Markdown parser

### ISSUE THREAD

**Issue by franklinyu** [MacDownApp/macdown#704]
Currently we are using master branch of Hoedown, which has a lot of bugs. There are (at least) two better alternatives:

1. Develop branch of Hoedown. I guess the API is mostly the same.
2. [cmark](https://github.com/jgm/cmark), the C reference implementation of CommonMark, currently under active development.

Both depends on CMake<sup>[1](#foot1-1)</sup>, so I am still worrying about adding another dependency. In addition, neither is accessible through CocoaPods; I am not sure how to link a Git submodule.

<sub><a name="foot1-1">1</a>: Actually [Homebrew](http://brew.sh/) user can avoid CMake by installing bottled cmark, but that's unfair to [MacPorts](https://www.macports.org/) users.</sub>

**Comment by ThePooE** [comment]
I'm adding this for the reference of the alternatives.
https://github.com/markdown/markdown.github.com/wiki/Implementations

**Comment by ijmitch** [comment]
I would be very interested to see Macdown fully support commonmark - especially the handling of <div> blocks with markdown inside.

**Comment by jdickey** [comment]
I added the +1 to @ijmitch's comment above, but I also wonder how many other users are using Macdown to author content for sites that do not fully support CommonMark, such as GitHub. I've been following CommonMark with bated breath for well over a year now, but I'm disappointed to see it apparently limited to content created and consumed within the same app/platform. I'd be interested in @franklinyu's and others' thoughts on this.

**Comment by franklinyu** [comment]
@jdickey Yes, the popularity of CommonMark is far from ideal. Currently my work-around is to use only the intersection of the behaviors. However, for MacDown, it has to choose a behavior; for example, GitHub Flavored Markdown is also a behavior (though perhaps not well-defined). What we can do right now is to choose one particular behavior before we cross our fingers and wait.

Among the available behaviors, I prefer CommonMark not because it's overwhelmingly popular (no one is, as far as I know), but because at least it's well-defined (or it tries to be well-defined).

> It's sad, so sad
> It's a sad, sad situation

--[*Sorry Seems To Be The Hardest Word*](http://www.azlyrics.com/lyrics/eltonjohn/sorryseemstobethehardestword.html), Elton John

**Comment by uranusjr** [comment]
The main thing stopping MacDown adopting CommonMark is the feature set. Unfortunately parsers other than Hoedown generally either lack a lot of non-standard (but popular) features, or don’t offer a good way for lib users to extend the parser, including all popular CommonMark parsers I looked into.

Please let me know, anyone, if you can find (or build!) a Markdown parser that suits the need!

**Comment by franklinyu** [comment]
Besides Remarkable mentioned in #723, [markdown-it](https://github.com/markdown-it/markdown-it) is another option.

**Comment by FranklinYu** [comment]
[marked](https://www.npmjs.com/package/marked) was mentioned in https://github.com/MacDownApp/macdown/issues/148#issuecomment-198147839.
